database2/ffboltdb/blockio.go

```diff
@@ -0,0 +1,746 @@
+//
+// The write cursor will also be advanced the number of bytes actually written
+// in the event of failure.
+//
+// Format: <network><block length><serialized block><checksum>
+func (s *blockStore) writeBlock(rawBlock []byte) (blockLocation, error) {
+	// Compute how many bytes will be written.
+	// 4 bytes each for block network + 4 bytes for block length +
+	// length of raw block + 4 bytes for checksum.
+	blockLen := uint32(len(rawBlock))
+	fullLen := blockLen + 12
+
+	// Move to the next block file if adding the new block would exceed the
+	// max allowed size for the current block file.  Also detect overflow
+	// to be paranoid, even though it isn't possible currently, numbers
+	// might change in the future to make possible.
+	//
+	// NOTE: The writeCursor.offset field isn't protected by the mutex
+	// since it's only read/changed in this function which can only be
+	// called during a write transaction, of which there can be only one at
+	// a time.
+	wc := s.writeCursor
+	finalOffset := wc.curOffset + fullLen
+	if finalOffset < wc.curOffset || finalOffset > s.maxBlockFileSize {
+		// This is done under the write cursor lock since the fileNum
+		// field is accessed elsewhere by readers.
+		//
+		// Close the current write file to force a read-only reopen
+		// with LRU tracking.  The close is done under the write lock
+		// for the file to prevent it from being closed out from under
+		// any readers currently reading from it.
+		wc.Lock()
+		wc.curFile.Lock()
+		if wc.curFile.file != nil {
+			_ = wc.curFile.file.Close()
+			wc.curFile.file = nil
+		}
+		wc.curFile.Unlock()
+
+		// Start writes into next file.
+		wc.curFileNum++
+		wc.curOffset = 0
+		wc.Unlock()
+	}
+
+	// All writes are done under the write lock for the file to ensure any
+	// readers are finished and blocked first.
+	wc.curFile.Lock()
+	defer wc.curFile.Unlock()
+
+	// Open the current file if needed.  This will typically only be the
+	// case when moving to the next file to write to or on initial database
+	// load.  However, it might also be the case if rollbacks happened after
+	// file writes started during a transaction commit.
+	if wc.curFile.file == nil {
+		file, err := s.openWriteFileFunc(wc.curFileNum)
+		if err != nil {
+			return blockLocation{}, err
+		}
+		wc.curFile.file = file
+	}
+
+	// Bitcoin network.
+	origOffset := wc.curOffset
+	hasher := crc32.New(castagnoli)
+	var scratch [4]byte
+	byteOrder.PutUint32(scratch[:], uint32(s.network))
+	n, err := wc.curFile.file.WriteAt(scratch[:], int64(wc.curOffset))
+	wc.curOffset += uint32(n)
+	if err != nil {
+		str := fmt.Sprintf("failed to write network to file %d at "+
+			"offset %d: %v", wc.curFileNum, wc.curOffset, err)
+		return blockLocation{}, makeDbErr(database.ErrDriverSpecific,
+			str, err)
+	}
+	_, _ = hasher.Write(scratch[:])
+
+	// Block length.
+	byteOrder.PutUint32(scratch[:], blockLen)
+	n, err = wc.curFile.file.WriteAt(scratch[:], int64(wc.curOffset))
+	wc.curOffset += uint32(n)
```

Comment:
It might be a little tricky with the mutexes.  Perhaps something like this?

```Go
// .... MUST be called with the write cursor current file lock held...
func (wc *writeCursor) Write(data []byte, fieldName string) error {
	n, err := wc.curFile.WriteAt(data, int64(wc.curOffset))
	wc.curOffset += uint32(n)
	if err != nil {
		str := fmt.Sprintf("failed to write %s to file %d at "+
			"offset %d: %v", fieldName, wc.curFileNum,
			wc.curOffset-n, err)
		return makeDbErr(database.ErrDriverSpecific, str, err)
	}
	return nil
}
```

Then all of these sections would just become:

```Go
if err := wc.Write(scratch[:], "network"); err != nil {
	return blockLocation{}, err
}
if err := wc.Write(scratch[:], "block length"); err != nil {
	return blockLocation{}, err
}
...
```